domains/client/domain-operator/src/fraud_proof.rs

```diff
@@ -532,39 +497,115 @@ where
+                        Vec::new(),
+                    )
+                }
+                (TraceDiffType::Mismatch, _) | (TraceDiffType::Shorter, _) => {
+                    let extrinsic_index = trace_mismatch_index as usize - 1;
+                    let target_extrinsic = encoded_extrinsics.get(extrinsic_index).ok_or(
+                        FraudProofError::OutOfBoundsExtrinsicIndex {
+                            index: extrinsic_index,
+                            max: max_extrinsic_index,
+                        },
+                    )?;
+                    let proof_of_inclusion = StorageProofProvider::<
+                        LayoutV1<<Block::Header as HeaderT>::Hashing>,
+                    >::generate_enumerated_proof_of_inclusion(
+                        encoded_extrinsics.as_slice(),
+                        extrinsic_index as u32,
+                    )
+                    .ok_or(FraudProofError::FailToGenerateProofOfInclusion)?;
+                    (
+                        ExecutionPhase::ApplyExtrinsic {
+                            extrinsic_proof: proof_of_inclusion,
+                            mismatch: if trace_diff_type == TraceDiffType::Mismatch {
+                                ApplyExtrinsicMismatch::StateRoot(trace_mismatch_index)
+                            } else {
+                                ApplyExtrinsicMismatch::Shorter
+                            },
+                        },
+                        block_builder.prepare_storage_changes_before(extrinsic_index)?,
+                        target_extrinsic.clone(),
+                    )
+                }
+            };
 
         let delta = storage_changes.transaction;
         let post_delta_root = storage_changes.transaction_storage_root;
         let execution_proof = prover.prove_execution(
             parent_header.hash(),
             &execution_phase,
-            target_extrinsic,
+            call_data.as_slice(),
             Some((delta, post_delta_root)),
         )?;
 
         Ok((execution_proof, execution_phase))
     }
 }
 
-/// Returns the index of first mismatch between the receipts `local` and `other` if any.
+/// Returns first mismatch (TraceDiffType, index) between the receipts `local` and `other` if any.
+/// If local trace length > other trace length then we provide storage proof as usual but add a flag in fraud proof
+/// indicating that this is length mismatch, so we create a storage proof with ApplyExtrinsic execution phase
+/// and prove that this state transition is valid, that means execution should not stop here.
+///
+/// if other trace length > local trace length then we create a storage proof with FinalizeBlock execution phase
+/// to prove that it is valid, so that means execution should stop here.
 pub(crate) fn find_trace_mismatch<Hash: Copy + Eq>(
     local_trace: &[Hash],
     other_trace: &[Hash],
-) -> Option<u32> {
-    local_trace
+) -> Option<(TraceDiffType, u32)> {
```

Comment:
The code may simplify a lot if `find_trace_mismatch` returns `Option<ExecutionPhase>` and `generate_invalid_state_transition_proof` generates the execution proof based on `ExecutionPhase` directly.